lib/service/connect.go

```diff
@@ -1041,55 +1047,99 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
-		process.log.Errorf("- connecting to auth server directly: %v", directErr)
-		if trace.IsConnectionProblem(err) && strings.Contains(err.Error(), "connection refused") {
-			err = trace.Wrap(err, "This is the alternative port we tried and it's not configured.")
+		directClient, directErr := connectToAuthServer(logger)
+		if directErr == nil {
+			return directClient, nil
 		}
-		process.log.Errorf("- connecting to auth server through tunnel: %v", err)
-		collectedErrs := trace.NewAggregate(directErr, err)
-		if utils.IsUntrustedCertErr(collectedErrs) {
-			collectedErrs = trace.WrapWithMessage(collectedErrs, utils.SelfSignedCertsMsg)
+
+		// Don't attempt to connect through a tunnel as a proxy or auth server.
+		if identity.ID.Role == types.RoleAuth || identity.ID.Role == types.RoleProxy {
+			return nil, trace.Wrap(directErr)
 		}
-		return nil, trace.WrapWithMessage(collectedErrs,
-			"Failed to connect to Auth Server directly or over tunnel, no methods remaining.")
+
+		// if that fails, attempt to connect to the auth server through a tunnel
+
+		logger.Debug("Attempting to discover reverse tunnel address.")
+		logger.Debug("Attempting to connect to Auth Server through tunnel.")
+
+		tunnelClient, err := process.newClientThroughTunnel(authServers, tlsConfig, sshClientConfig)
+		if err != nil {
+			process.log.Errorf("Node failed to establish connection to Teleport Proxy. We have tried the following endpoints:")
+			process.log.Errorf("- connecting to auth server directly: %v", directErr)
+			if trace.IsConnectionProblem(err) && strings.Contains(err.Error(), "connection refused") {
+				err = trace.Wrap(err, "This is the alternative port we tried and it's not configured.")
+			}
+			process.log.Errorf("- connecting to auth server through tunnel: %v", err)
+			collectedErrs := trace.NewAggregate(directErr, err)
+			if utils.IsUntrustedCertErr(collectedErrs) {
+				collectedErrs = trace.WrapWithMessage(collectedErrs, utils.SelfSignedCertsMsg)
+			}
+			return nil, trace.WrapWithMessage(collectedErrs,
+				"Failed to connect to Auth Server directly or over tunnel, no methods remaining.")
+		}
+
+		logger.Debug("Connected to Auth Server through tunnel.")
+		return tunnelClient, nil
+
+	// for config v3, either tunnel to the given proxy server or directly connect to the given auth server
+	case defaults.TeleportConfigVersionV3:
+		proxyServer := process.Config.ProxyServer
+		if !proxyServer.IsEmpty() {
+			logger := process.log.WithField("proxy-server", proxyServer.String())
+			logger.Debug("Attempting to connect to Proxy Server through tunnel.")
+
+			tunnelClient, err := process.newClientThroughTunnel([]utils.NetAddr{proxyServer}, tlsConfig, sshClientConfig)
+			if err != nil {
+				return nil, trace.Errorf("Failed to connect to Proxy Server through tunnel: %v", err)
+			}
+
+			logger.Debug("Connected to Proxy Server through tunnel.")
```

Comment:
I think this would be more accurate, the proxy provides the tunnel but you are connecting to the auth server
```suggestion
			logger.Debug("Connected to Auth Server through tunnel.")
```

Reply:
Yeah that makes sense